Bottom: 21a3eb4fa6542dfd936115f7f3d204d736b581e2
Top:    71db3a1059f6bb78e6a8bbd4e52210b344c9002d
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2015-04-21 23:29:32 +0100

Refresh of spent-all-that-time-fixing

---

diff --git a/src/hid/step/step.c b/src/hid/step/step.c
index 8807835..2619ce8 100644
--- a/src/hid/step/step.c
+++ b/src/hid/step/step.c
@@ -867,6 +867,8 @@ step_do_export (HID_Attr_Val * options)
         double ax, ay, az;
         double rx, ry, rz;
         double rotation;
+        double cos_rot;
+        double sin_rot;
         GList *model_iter;
 
         /* Skip if the component doesn't have a STEP-AP214 3d_model */
@@ -940,16 +942,31 @@ step_do_export (HID_Attr_Val * options)
             models = g_list_append (models, model);
           }
 
+        cos_rot = cos (rotation * M_PI / 180.);
+        sin_rot = sin (rotation * M_PI / 180.);
+
+        // Rotation of part on board
+        // (NB: Y flipped from normal right handed convention)
+        //[cos -sin   0] [x] = [xcos - ysin]
+        //[sin  cos   0] [y]   [xsin + ycos]
+        //[  0    0   1] [z]   [z          ]
+
+        // Flip of part to backside of board
+        // [  1   0   0] [x] = [ x]
+        // [  0  -1   0] [y] = [-y]
+        // [  0   0  -1] [z] = [-z]
+
         instance = g_new0 (struct assembly_model_instance, 1);
         instance->name = NAMEONPCB_NAME (element);
-        instance->ox = COORD_TO_MM (element->MarkX) + ox,
-        instance->oy = COORD_TO_MM (element->MarkY) + oy,
-        instance->oz = on_solder_negate * -COORD_TO_MM (HACK_BOARD_THICKNESS) / 2.0 + oz; /* XXX: THINK ABOUT HOW WE FLIP FOR NON-TRIVIAL STARING TRANSFORMS */
-        instance->ax = on_solder_negate * ax,                       /* XXX: THINK ABOUT HOW WE FLIP TO THE BACKSIDE FOR NON-TRIVIAL STARING TRANSFORMS */
-        instance->ay = on_solder_negate * ay,                       /* XXX: ALSO THINK ABOUT HOW WE ROTATE THE COMPONENT BY ANGLE "rotation"           */
+        instance->ox = COORD_TO_MM (element->MarkX) +                    ( ox * cos_rot + oy * sin_rot),
+        instance->oy = COORD_TO_MM (element->MarkY) + on_solder_negate * (-ox * sin_rot + oy * cos_rot),
+        instance->oz = on_solder_negate * -COORD_TO_MM (HACK_BOARD_THICKNESS) / 2.0 + on_solder_negate * oz;
+        instance->ax =                    ( ax * cos_rot + ay * sin_rot),
+        instance->ay = on_solder_negate * (-ax * sin_rot + ay * cos_rot),
         instance->az = on_solder_negate * az;
-        instance->rx = rx, instance->ry = ry,  instance->rz = rz; /* XXX: THINK ABOUT HOW WE FLIP TO THE BACKSIDE FOR NON-TRIVIAL STARING TRANSFORMS */
-                                                                   /* XXX: ALSO THINK ABOUT HOW WE ROTATE THE COMPONENT BY ANGLE "rotation"           */
+        instance->rx =                    ( rx * cos_rot + ry * sin_rot), /* XXX: SHOULD THIS FACTOR IN on_solder_negate? */
+        instance->ry = on_solder_negate * (-rx * sin_rot + ry * cos_rot), /* XXX: SHOULD THIS FACTOR IN ol_solder_negaet? */
+        instance->rz = on_solder_negate * rz;
         model->instances = g_list_append (model->instances, instance);
 
       }
